### PR TITLE
MCOL-4171

### PR DESCRIPTION
**Pull request by davidhall586**
Window Function changes for decimal(38)

### CODE REVIEW

File: dbcon/execplan/windowfunctioncolumn.cpp [hunk 1]
```diff
@@ -387,9 +389,16 @@ void WindowFunctionColumn::adjustResultType()
         boost::iequals(fFunctionName, "AVG") ||
         boost::iequals(fFunctionName, "AVG_DISTINCT"))
     {
-        fResultType.colDataType = CalpontSystemCatalog::LONGDOUBLE;
-        fResultType.colWidth = sizeof(long double);
-        fResultType.precision = -1;
+        if (fFunctionParms[0]->resultType().colDataType == CalpontSystemCatalog::DECIMAL)
```
Comment (tntnatbry) [inline dbcon/execplan/windowfunctioncolumn.cpp]: Don't we also need to check for UDECIMAL here? As well as in the earlier if condition in this function?

File: dbcon/execplan/windowfunctioncolumn.cpp [hunk 2]
```diff
@@ -661,18 +670,35 @@ void WindowFunctionColumn::evaluate(Row& row, bool& isNull)
...
-                default:
+                case 8:
                 {
                     if (row.equals<8>(BIGINTNULL, fInputIndex))
                         isNull = true;
                     else
                     {
-                        fResult.decimalVal.value = (int64_t)row.getUintField<8>(fInputIndex);
+                        fResult.decimalVal.value = row.getIntField<8>(fInputIndex);
+                        fResult.decimalVal.scale = (unsigned)fResultType.scale;
```
Comment (tntnatbry) [inline dbcon/execplan/windowfunctioncolumn.cpp:683]: Don't we also need to set fResult.decimalVal.precision here?

File: utils/messageqcpp/bytestream.cpp [hunk 3]
```diff
@@ -236,6 +236,17 @@ ByteStream& ByteStream::operator<<(const uint64_t o)
 }
 
 // WIP MCOL-641
+ByteStream& ByteStream::operator<<(const int128_t& o)
```
Comment (tntnatbry) [inline utils/messageqcpp/bytestream.cpp]: @drrtuy has already added support for int128_t in bytestream.h/.cpp in his PR: https://github.com/mariadb-corporation/mariadb-columnstore-engine/pull/1346/files

Can you undo this change?

File: utils/regr/moda.h [hunk 4]
```diff
@@ -54,6 +55,38 @@
...
+struct hasher
+{
+    inline size_t operator()(T val) const
+    {
+        return fHasher((char*) &val, sizeof(T));
+    }
+
+private:
+    utils::Hasher fHasher;
+};
```
Comment (tntnatbry) [inline utils/regr/moda.h:69]: I wasn't aware of this usage of the Hasher class from hasher.h. In the casual partitioning PR, I added Hash128 and Equal128 classes in hasher.h: https://github.com/mariadb-corporation/mariadb-columnstore-engine/pull/1349/files#diff-2f5b44d849a51ebb552b1793a8ed93cdR369

As you will see, I am merely returning the lower 64 bytes as the hash value, which does not look like the right approach, as it will lead to all 2^64 values in the upper 64 bytes getting the same hash values for the same bit pattern in the lower 64 bytes -> significant collisions.

We should consolidate this and just have this implementation at a single place.

File: utils/rowgroup/rowgroup.h [hunk 5]
```diff
@@ -343,14 +343,17 @@ class Row
     template<int len> inline bool equals(uint64_t val, uint32_t colIndex) const;
     inline bool equals(long double val, uint32_t colIndex) const;
     bool equals(const std::string& val, uint32_t colIndex) const;
-
+    inline bool equals(int128_t val, uint32_t colIndex) const;
```
Comment (tntnatbry) [inline utils/rowgroup/rowgroup.h]: Shouldn't val be a const ref?

File: utils/rowgroup/rowgroup.h [hunk 6]
```diff
@@ -380,6 +383,7 @@ class Row
     inline void setFloatField(float val, uint32_t colIndex);
     inline void setDecimalField(double val, uint32_t colIndex) { };  // TODO: Do something here
     inline void setLongDoubleField(long double val, uint32_t colIndex);
+    inline void setInt128Field(int128_t val, uint32_t colIndex);
```
Comment (tntnatbry) [inline utils/rowgroup/rowgroup.h]: Shouldn't val be a const ref?

File: utils/windowfunction/wf_percentile.cpp [hunk 7]
```diff
@@ -90,6 +88,32 @@ boost::shared_ptr<WindowFunctionType> WF_percentile<T>::makeFunction(int id, con
                 break;
             }
 
+            case CalpontSystemCatalog::DECIMAL:
+            {
+                if (wc->functionParms()[0]->resultType().colWidth < 16)
+                {
+                    func.reset(new WF_percentile<int128_t>(id, name));
```
Comment (tntnatbry) [inline utils/windowfunction/wf_percentile.cpp]: Should be int64_t here.

File: utils/windowfunction/wf_percentile.cpp [hunk 8]
```diff
@@ -90,6 +88,32 @@ boost::shared_ptr<WindowFunctionType> WF_percentile<T>::makeFunction(int id, con
...
+                }
+                break;
+            }
+            
+            case CalpontSystemCatalog::UDECIMAL:
+            {
+                if (wc->functionParms()[0]->resultType().colWidth < 16)
+                {
+                    func.reset(new WF_percentile<uint128_t>(id, name));
```
Comment (tntnatbry) [inline utils/windowfunction/wf_percentile.cpp]: Should be uint64_t here.

File: utils/windowfunction/wf_sum_avg.cpp [hunk 9]
```diff
@@ -1,4 +1,4 @@
-/* Copyright (C) 2014 InfiniDB, Inc.
+/* CopyrighT (C) 2014 InfiniDB, Inc.
```
Comment (tntnatbry) [inline utils/windowfunction/wf_sum_avg.cpp:1]: The lower case 't' looked perfectly ok earlier :)

File: utils/windowfunction/wf_sum_avg.cpp [hunk 10]
```diff
@@ -50,112 +50,116 @@ using namespace joblist;
...
-
-long double avgWithLimit(long double sum, uint64_t count, int scale, long double u, long double l)
-{
-    long double factor = pow(10.0, scale);
-    long double avg = sum / count;
-    avg *= factor;
-    avg += (avg < 0) ? (-0.5) : (0.5);
-
-    if (avg > u || avg < l)
+    int128_t factor = pow(10.0, scale);
```
Comment (tntnatbry) [inline utils/windowfunction/wf_sum_avg.cpp]: I don't think the pow() library function supports int128_t. You might want to look at how we are calculating this factor in some of the utils/funcexp support that we added. I am posting a code snippet from Func_ceil::getUintVal for easy reference (see the usage of datatypes::getScaleDivisor(scaleDivisor, d.scale);):

```
        case execplan::CalpontSystemCatalog::DECIMAL:
        case execplan::CalpontSystemCatalog::UDECIMAL:
        {
            IDB_Decimal d = parm[0]->data()->getDecimalVal(row, isNull);

            if (isNull)
                break;

            // negative scale is not supported by CNX yet
            if (d.scale > 0)
            {
                if (d.scale > datatypes::INT128MAXPRECISION)
                {
                    std::ostringstream oss;
                    oss << "ceil: datatype of " << execplan::colDataTypeToString(op_ct.colDataType)
                        << " with scale " << (int) d.scale << " is beyond supported scale";
                    throw logging::IDBExcept(oss.str(), ERR_DATATYPE_NOT_SUPPORT);
                }

                if (op_ct.colWidth == datatypes::MAXDECIMALWIDTH)
                {
                    int128_t tmp = d.s128Value;
                    int128_t scaleDivisor;
                    datatypes::getScaleDivisor(scaleDivisor, d.scale);
                    d.s128Value /= scaleDivisor;

                    // Add 1 if this is a positive number and there were values to the right of the
                    // decimal point so that we return the largest integer value not less than X.
                    if ((tmp - (d.s128Value * scaleDivisor)) > 0)
                        d.s128Value += 1;

                    ret = datatypes::Decimal::getUInt64FromWideDecimal(d.s128Value);
                }
                else
                {
                    int64_t tmp = d.value;
                    d.value /= helpers::powerOf10_c[d.scale];

                    // Add 1 if this is a positive number and there were values to the right of the
                    // decimal point so that we return the largest integer value not less than X.
                    if ((tmp - (d.value * helpers::powerOf10_c[d.scale])) > 0)
                        d.value += 1;

                    ret = (uint64_t) d.value;
                }
            }
        }
        break;
```

File: utils/windowfunction/wf_sum_avg.cpp [hunk 11]
```diff
@@ -50,112 +50,116 @@ using namespace joblist;
...
-
-template<>
-int64_t calculateAvg<int64_t>(int64_t sum, uint64_t count, int scale)
+template<typename T_IN, typename T_OUT>
+uint128_t WF_sum_avg<T_IN, T_OUT>::calculateAvg(uint128_t sum, uint64_t count, int scale)
 {
-    int64_t t = (int64_t) avgWithLimit(sum, count, scale,
-                                       numeric_limits<int64_t>::max(), numeric_limits<int64_t>::min());
-    return t;
+    uint128_t factor = pow(10.0, scale);
```
Comment (tntnatbry) [inline utils/windowfunction/wf_sum_avg.cpp]: Same here. See my comment above.

File: utils/windowfunction/wf_udaf.cpp [hunk 12]
```diff
@@ -542,11 +542,13 @@ void WF_udaf::SetUDAFValue(static_any::any& valOut, int64_t colOut,
     static const static_any::any& intTypeId = (int)1;
     static const static_any::any& longTypeId = (long)1;
     static const static_any::any& llTypeId = (long long)1;
+    static const static_any::any& int128TypeId = (int128_t)1;
     static const static_any::any& ucharTypeId = (unsigned char)1;
     static const static_any::any& ushortTypeId = (unsigned short)1;
     static const static_any::any& uintTypeId = (unsigned int)1;
     static const static_any::any& ulongTypeId = (unsigned long)1;
     static const static_any::any& ullTypeId = (unsigned long long)1;
+    static const static_any::any& uint128TypeId = (int128_t)1;
```
Comment (tntnatbry) [inline utils/windowfunction/wf_udaf.cpp]: This should be uint128_t?

File: utils/windowfunction/windowfunctiontype.cpp [hunk 13]
```diff
@@ -337,6 +357,11 @@ template<> void WindowFunctionType::setValue<long double>(uint64_t i, long doubl
     fRow.setLongDoubleField(t, i);
 }
 
+template<> void WindowFunctionType::setValue<int128_t>(uint64_t i, int128_t& t)
+{
+    fRow.setInt128Field(t, i);
+}
```
Comment (tntnatbry) [inline utils/windowfunction/windowfunctiontype.cpp:363]: Shouldn't t be a const ref here? And how about specialization of setValue for uint128_t?

File: utils/windowfunction/windowfunctiontype.cpp [hunk 14]
```diff
@@ -483,6 +518,40 @@ void WindowFunctionType::implicit2T(uint64_t i, T& t, int s)
...
         }
 
+        case CalpontSystemCatalog::DECIMAL:
+        {
+            uint32_t w = fRow.getColumnWidth(i);
+            if (w < 16)
+                t = (T) fRow.getIntField(i);
+            else 
+                t = (T) fRow.getInt128Field(i);
+            pw = s - fRow.getScale(i); // pw is difference of scales, will be in [-18, 18]
```
Comment (tntnatbry) [inline utils/windowfunction/windowfunctiontype.cpp]: [-38, 38] ?

File: utils/windowfunction/windowfunctiontype.cpp [hunk 15]
```diff
@@ -483,6 +518,40 @@ void WindowFunctionType::implicit2T(uint64_t i, T& t, int s)
...
+
+            if (pw > 0)
+                t *= IDB_pow[pw];
```
Comment (tntnatbry) [inline utils/windowfunction/windowfunctiontype.cpp]: IDB_pow does not handle int128_t if pw > 18. Refer to one of my earlier comments regarding this.

File: utils/windowfunction/windowfunctiontype.cpp [hunk 16]
```diff
@@ -483,6 +518,40 @@ void WindowFunctionType::implicit2T(uint64_t i, T& t, int s)
...
+        }
+
+        case CalpontSystemCatalog::UDECIMAL:
+        {
+            uint32_t w = fRow.getColumnWidth(i);
+            if (w < 16)
+                t = (T) fRow.getUintField(i);
+            else 
+                t = (T) fRow.getUint128Field(i);
+            pw = s - fRow.getScale(i); // pw is difference of scales, will be in [-18, 18]
```
Comment (tntnatbry) [inline utils/windowfunction/windowfunctiontype.cpp]: [-38, 38] ?

File: utils/windowfunction/windowfunctiontype.cpp [hunk 17]
```diff
@@ -483,6 +518,40 @@ void WindowFunctionType::implicit2T(uint64_t i, T& t, int s)
...
+
+            if (pw > 0)
+                t *= IDB_pow[pw];
```
Comment (tntnatbry) [inline utils/windowfunction/windowfunctiontype.cpp]: IDB_pow does not handle int128_t if pw > 18. Refer to one of my earlier comments regarding this.

File: utils/windowfunction/windowfunctiontype.h [hunk 18]
```diff
@@ -238,6 +242,10 @@ class WindowFunctionType
     {
         fRow.setIntField(v, i);
     }
+    void    setInt128Value(int64_t i, int128_t v)
```
Comment (tntnatbry) [inline utils/windowfunction/windowfunctiontype.h]: Shouldn't v be a const ref here?

File: writeengine/wrapper/writeengine.cpp [hunk 19]
```diff
@@ -837,7 +837,18 @@ int WriteEngineWrapper::deleteRow(const TxnID& txnid, const vector<CSCTypesList>
...
+        case WriteEngine::WR_VARBINARY : // treat same as char for now
+        case WriteEngine::WR_CHAR:
+        case WriteEngine::WR_BLOB:
+        case WriteEngine::WR_TEXT:
+            valArray = (char*) calloc(sizeof(char), totalRow * MAX_COLUMN_BOUNDARY);
+            break;
+        default:
+            valArray = calloc(totalRow, colWidth);
+            break;
+    }
```
Comment (tntnatbry) [inline writeengine/wrapper/writeengine.cpp:854]: For WriteEngine::WR_TOKEN, is colWidth == sizeof(Token) ?

**Review by tntnatbry — CHANGES_REQUESTED** [review]

File: dbcon/execplan/windowfunctioncolumn.cpp (hunk 1 above)
Comment (davidhall586) [reply dbcon/execplan/windowfunctioncolumn.cpp]: Yes

File: utils/messageqcpp/bytestream.cpp (hunk 3 above)
Comment (davidhall586) [reply utils/messageqcpp/bytestream.cpp]: Yes

File: utils/rowgroup/rowgroup.h (hunk 5 above)
Comment (davidhall586) [reply utils/rowgroup/rowgroup.h]: It could be. I don't think it matters very much. If you like that better, I'll change it. It might save one mv assembly code, depending on the micro code implementation.

File: utils/windowfunction/windowfunctiontype.cpp (hunk 15 above)
Comment (davidhall586) [reply utils/windowfunction/windowfunctiontype.cpp]: I don't see an earlier comment. Should we resize IDB_pow, or use std::pow()?

File: utils/windowfunction/windowfunctiontype.cpp (hunk 16 above)
Comment (davidhall586) [reply utils/windowfunction/windowfunctiontype.cpp]: Lol. Of course I didn't think to update the comment.

File: writeengine/wrapper/writeengine.cpp (hunk 19 above)
Comment (davidhall586) [reply writeengine/wrapper/writeengine.cpp:854]: No, colwidth is the size specified in the SQL. CHAR(1) will have a colwidth 1. CHAR(28) will have a colwidth 28. Hence the hard coded Constant "8"

File: dbcon/execplan/windowfunctioncolumn.cpp (hunk 2 above)
Comment (davidhall586) [reply dbcon/execplan/windowfunctioncolumn.cpp:683]: No. output precision is not specified. We don't want to limit the output precision of aggregated values. It could lead to all sorts of unintended truncation. The default precision is 0. This is the way it has worked for many years.

File: utils/regr/moda.h (hunk 4 above)
Comment (davidhall586) [reply utils/regr/moda.h:69]: Here's another implementation of it: TupleJoiner::hasher. This one doesn't handle int128 at all.

File: utils/messageqcpp/bytestream.cpp (hunk 3 above)
Comment (davidhall586) [reply utils/messageqcpp/bytestream.cpp]: Done

File: utils/rowgroup/rowgroup.h (hunk 5 above)
Comment (davidhall586) [reply utils/rowgroup/rowgroup.h]: I wet ahead and made these const ref, since it will save a couple of pico seconds

File: utils/windowfunction/windowfunctiontype.cpp (hunk 15 above)
Comment (davidhall586) [reply utils/windowfunction/windowfunctiontype.cpp]: So, I created another constant table for int128 > 18 digits and everybody calls datatypes::getScaleDivisor.

File: utils/windowfunction/windowfunctiontype.cpp (hunk 16 above)
Comment (davidhall586) [reply utils/windowfunction/windowfunctiontype.cpp]: This code is now obsolete, having been replaced by a call to datatypes::getScaleDivisor()

File: utils/windowfunction/windowfunctiontype.cpp (hunk 17 above)
Comment (davidhall586) [reply utils/windowfunction/windowfunctiontype.cpp]: Replaced by datatypes::getScaleDivisor()

File: utils/windowfunction/windowfunctiontype.h (hunk 18 above)
Comment (davidhall586) [reply utils/windowfunction/windowfunctiontype.h]: Changed

File: utils/regr/moda.h (hunk 4 above)
Comment (davidhall586) [reply utils/regr/moda.h:69]: I've created a MCOL to consolidate the hashers into one place. I believe most can be done with the template class I created for moda.

File: dbcon/execplan/windowfunctioncolumn.cpp (hunk 1 above)
Comment (davidhall586) [reply dbcon/execplan/windowfunctioncolumn.cpp]: Done

File: datatypes/mcs_decimal.h [hunk 20]
```diff
@@ -73,21 +136,21 @@ constexpr int128_t Decimal128Null = int128_t(0x8000000000000000LL) << 64;
...
-            divisor *= mcs_pow_10[maxPowOf10];
-            divisor *= mcs_pow_10[maxPowOf10];
-            break;
-        case 1:
-            divisor *= mcs_pow_10[maxPowOf10];
-        case 0:
-            divisor *= mcs_pow_10[scale%maxPowOf10];
-        default:
-            break;
+        divisor = mcs_pow_10_128[scale-18];
```
Comment (tntnatbry) [inline datatypes/mcs_decimal.h]: Shouldn't this be `scale-19`?

File: utils/rowgroup/rowgroup.h (hunk 6 above)
Comment (tntnatbry) [reply utils/rowgroup/rowgroup.h]: @dhall-MariaDB const ref here as well, just as you did in the equals() above?

File: utils/windowfunction/wf_percentile.cpp (hunk 7 above)
Comment (tntnatbry) [reply utils/windowfunction/wf_percentile.cpp]: @dhall-MariaDB ^^

File: utils/windowfunction/wf_percentile.cpp (hunk 8 above)
Comment (tntnatbry) [reply utils/windowfunction/wf_percentile.cpp]: @dhall-MariaDB ^^

File: utils/windowfunction/wf_sum_avg.cpp (hunk 10 above)
Comment (tntnatbry) [reply utils/windowfunction/wf_sum_avg.cpp]: @dhall-MariaDB ^^ . Now that you have added mcs_pow_10_128. That can be used here.

File: utils/windowfunction/wf_sum_avg.cpp (hunk 11 above)
Comment (tntnatbry) [reply utils/windowfunction/wf_sum_avg.cpp]: @dhall-MariaDB ^^

File: utils/windowfunction/wf_udaf.cpp (hunk 12 above)
Comment (tntnatbry) [reply utils/windowfunction/wf_udaf.cpp]: @dhall-MariaDB ^^

File: utils/windowfunction/windowfunctiontype.cpp (hunk 13 above)
Comment (tntnatbry) [reply utils/windowfunction/windowfunctiontype.cpp:363]: @dhall-MariaDB ^^

File: utils/windowfunction/windowfunctiontype.h (hunk 18 above)
Comment (tntnatbry) [reply utils/windowfunction/windowfunctiontype.h]: I don't see this change.

File: writeengine/wrapper/writeengine.cpp (hunk 19 above)
Comment (tntnatbry) [reply writeengine/wrapper/writeengine.cpp:854]: I am a bit confused. I am referring to the following:

```
        case WriteEngine::WR_TOKEN:
            valArray = (Token*) calloc(sizeof(Token), totalRow);
            break;
```

This is now replaced by the default case: `valArray = calloc(totalRow, colWidth);`

Is this correct?

File: utils/windowfunction/windowfunctiontype.cpp [hunk 21]
```diff
@@ -480,7 +480,7 @@ template<typename T>
 void WindowFunctionType::implicit2T(uint64_t i, T& t, int s)
 {
     int ct = fRow.getColType(i);
-    int pw = 0;
+    int64_t divisor = 1;
```
Comment (tntnatbry) [inline utils/windowfunction/windowfunctiontype.cpp]: This should be int128_t?

File: utils/windowfunction/windowfunctiontype.cpp [hunk 22]
```diff
@@ -596,6 +556,15 @@ void WindowFunctionType::implicit2T(uint64_t i, T& t, int s)
             break;
         }
     }
+
+    s -= fRow.getScale(i); // we scale only the difference of scales
```
Comment (tntnatbry) [inline utils/windowfunction/windowfunctiontype.cpp:581]: For DOUBLE/UDOUBLE/FLOAT/UFLOAT/LONGDOUBLE, we didn't have `s -= fRow.getScale(i);` in the individual case statements above. Is this still OK for these datatypes?

**Review by tntnatbry — CHANGES_REQUESTED** [review]

File: utils/rowgroup/rowgroup.h (hunk 6 above)
Comment (davidhall586) [reply utils/rowgroup/rowgroup.h]: Didn't get pushed.

File: utils/windowfunction/wf_sum_avg.cpp (hunk 9 above)
Comment (davidhall586) [reply utils/windowfunction/wf_sum_avg.cpp:1]: LOL

Comment (davidhall586) [reply utils/windowfunction/wf_sum_avg.cpp:1]: Reminds me to update the copyright notices

File: utils/windowfunction/wf_sum_avg.cpp (hunk 10 above)
Comment (davidhall586) [reply utils/windowfunction/wf_sum_avg.cpp]: Yeah. I missed that one

File: utils/windowfunction/wf_sum_avg.cpp (hunk 11 above)
Comment (davidhall586) [reply utils/windowfunction/wf_sum_avg.cpp]: Same

File: utils/windowfunction/wf_udaf.cpp (hunk 12 above)
Comment (davidhall586) [reply utils/windowfunction/wf_udaf.cpp]: Fixed

File: utils/windowfunction/windowfunctiontype.cpp (hunk 13 above)
Comment (davidhall586) [reply utils/windowfunction/windowfunctiontype.cpp:363]: It probably should be const, but since the parent template isn't const ref, this can't be unless we go back and change a bunch. It's probably not worth the effort. const only exists to protect programmers from programmers. The machine don't care.

Comment (davidhall586) [reply utils/windowfunction/windowfunctiontype.cpp:363]: Added uint128_t stuff. Required to add to Row as well.

File: utils/windowfunction/windowfunctiontype.h (hunk 18 above)
Comment (davidhall586) [reply utils/windowfunction/windowfunctiontype.h]: Huh. I have it. Maybe it made it in this time.

File: writeengine/wrapper/writeengine.cpp (hunk 19 above)
Comment (davidhall586) [reply writeengine/wrapper/writeengine.cpp:854]: Sorry. I didn't parse your question. Now that I understand the question, the answer is: I can't definitely say. So I'll add token back into the case.

File: utils/windowfunction/windowfunctiontype.cpp (hunk 21 above)
Comment (davidhall586) [reply utils/windowfunction/windowfunctiontype.cpp]: It should b T

File: datatypes/mcs_decimal.h (hunk 20 above)
Comment (davidhall586) [reply datatypes/mcs_decimal.h]: Yes. And after some thought, I changed msc_pow-10 and mcs_pow_10_128 to signed, as it was in IDB_pow. Because getScaleDivisor is template based, it is most often a signed T. With the way it works, we still get a maximum scale of 38.

File: utils/windowfunction/windowfunctiontype.cpp (hunk 22 above)
Comment (davidhall586) [reply utils/windowfunction/windowfunctiontype.cpp:581]: The old code for double:
            if (s == 0)
                t = (T) fRow.getDoubleField(i);
            else
                t = (T) (fRow.getDoubleField(i) * IDB_pow[s]); // s is scale, [0, 18]
The only difference here is that scale is assumed to be positive because, for some reason, we're not taking the diff.
That being said, the only time s is non zero is when getting the default for LEAD|LAG (if not a constant). Currently, the MDB parser doesn't parse the default parameter for LEAD|LAG, so there's no way to test it. This is sleeping code -- unreachable -- until MDB  implements the full standard for Window Functions. Nonetheless, I changed a line in LEAD|LAG that I think looks wrong.

**Review by tntnatbry — APPROVED** [review]
